File: app/src/test/java/com/baro/helpers/FileHelperTest.java
package com.baro.helpers;

import android.graphics.Bitmap;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Scanner;

import mockit.Capturing;
import mockit.Expectations;
import mockit.Mocked;
import mockit.Tested;
import mockit.Verifications;
import mockit.integration.junit4.JMockit;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(JMockit.class)
public class FileHelperTest {


    @Test
    public void getFileAtPathShouldReturnFileIfFileExists(@Mocked File file) {
        Path pathToFile = Paths.get("path/to/parent", "filename");

        new Expectations() {{
            new File(pathToFile.toString());
            file.exists(); result = true;
            file.getAbsolutePath(); result = "path/to/parent/filename";
        }};


        File outputFile = FileHelper.getFileAtPath(pathToFile);
        assertEquals(outputFile.getAbsolutePath(), "path/to/parent/filename");
    }

    @Test
    public void getFileAtPathShouldReturnNullIfFileDoesNotExists(@Mocked File file) {
        Path pathToFile = Paths.get("path/to/nothing", "filename");

        new Expectations() {{
            new File(pathToFile.toString());
            file.exists(); result = false;
        }};

        File outputFile = FileHelper.getFileAtPath(pathToFile);
        assertNull(outputFile);
    }

    @Test
    public void createFileAtPathShouldReturnFileAtPathIfFileExists(@Mocked File file) {
        Path pathToFile = Paths.get("path/to/parent", "filename");

        new Expectations() {{
            new File(pathToFile.toString());
            file.exists(); result = true;
            file.getAbsolutePath(); result = "path/to/parent/filename";
        }};

        File outputFile = FileHelper.createFileAtPath(pathToFile);
        assertEquals(outputFile.getAbsolutePath(), "path/to/parent/filename");
    }


    @Test
    public void createDirectoryShouldReturnTrueIfDirectoryIsCreated() {
        File parent = new File("path/to/parent/directory");
        String directoryName = "Directory Name";

        assertTrue(FileHelper.createDirectory(parent, directoryName));
    }




    @Test
    public void writeToFileShouldReturnTrueIfTextWasWrittenToValidFilePath(@Mocked FileWriter fileWriter) throws IOException {
        String data = "Data";
        File file = new File("valid/path/to/file");

        new Expectations() {{
            new FileWriter(file);

            fileWriter.write(data);
            fileWriter.flush();
            fileWriter.close();
        }};

        boolean resultWritingFile = FileHelper.writeToFile(file, data);
        assertTrue(resultWritingFile);

    }

    @Test
    public void writeToFileShouldReturnFalseIfTextWasNotWrittenToInvalidFilePath(@Mocked FileWriter fileWriter) throws IOException {
        String data = "Data";
        File file = new File("invalid/path/to/file");

        new Expectations() {{
            new FileWriter(file); result = new IOException();
        }};

        boolean resultWritingFile = FileHelper.writeToFile(file, data);
        assertFalse(resultWritingFile);

    }

    @Test
    public void readFileShouldReturnFileContentWhenValidFilePath(@Mocked Scanner scanner) throws IOException {
        String data = "Data";
        File file = new File("valid/path/to/file");

        new Expectations() {{
            new Scanner(file);

            scanner.useDelimiter("\\Z"); result = scanner;
            scanner.next(); result = data;
        }};


        assertEquals(FileHelper.readFile(file), "Data");
    }


    @Test
    public void readFileShouldReturnNullWhenInvalidFilePath(@Mocked Scanner scanner) throws IOException {
        String data = "Data";
        File file = new File("invalid/path/to/file");

        new Expectations() {{
            new Scanner(file); result = new FileNotFoundException();
        }};

        assertNull(FileHelper.readFile(file));
    }


    @Test
    public void writeBitmapToFileShouldReturnTrueWhenBitmapIsSavedToValidFilePath(
            @Mocked FileOutputStream fos,
            @Mocked Bitmap bitmap) throws FileNotFoundException {

        File file = new File("valid/path/to/file");

        new Expectations() {{
            bitmap.compress(Bitmap.CompressFormat.JPEG, 100, new FileOutputStream(file)); result = true;
        }};
        assertTrue(FileHelper.writeBitmapToFile(file, bitmap));
    }

    @Test
    public void writeBitmapToFileShouldReturnFalseWhenBitmapIsNotSavedToInvalidFilePath(
            @Mocked FileOutputStream fos,
            @Mocked Bitmap bitmap) throws FileNotFoundException {

        File file = new File("invalid/path/to/file");

        new Expectations() {{
            new FileOutputStream(file); result = new FileNotFoundException();
        }};

        assertFalse(FileHelper.writeBitmapToFile(file, bitmap));
    }
}

File: app/src/test/java/com/baro/helpers/FileHelperTest.kt
package com.baro.helpers

import android.content.ContentResolver
import android.graphics.Bitmap
import android.net.Uri
import mockit.Expectations
import mockit.Mocked
import mockit.integration.junit4.JMockit
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import java.io.*
import java.nio.file.Paths
import java.util.*

@RunWith(JMockit::class)
class FileHelperTest {
    @Test
    fun getFileAtPathShouldReturnFileIfFileExists(@Mocked file: File?) {
        val pathToFile = Paths.get("path/to/parent", "filename")
        object : Expectations() {
            init {
                File(pathToFile.toString())
                file!!.exists()
                result = true
                file.absolutePath
                result = "path/to/parent/filename"
            }
        }
        val outputFile = FileHelper.getFileAtPath(pathToFile)
        Assert.assertEquals(outputFile?.absolutePath, "path/to/parent/filename")
    }

    @Test
    fun getFileAtPathShouldReturnNullIfFileDoesNotExists(@Mocked file: File?) {
        val pathToFile = Paths.get("path/to/nothing", "filename")
        object : Expectations() {
            init {
                File(pathToFile.toString())
                file!!.exists()
                result = false
            }
        }
        val outputFile = FileHelper.getFileAtPath(pathToFile)
        Assert.assertNull(outputFile)
    }

    @Test
    fun createFileAtPathShouldReturnFileAtPathIfFileExists(@Mocked file: File?) {
        val pathToFile = Paths.get("path/to/parent", "filename")
        object : Expectations() {
            init {
                File(pathToFile.toString())
                file!!.exists()
                result = true
                file.absolutePath
                result = "path/to/parent/filename"
            }
        }
        val outputFile = FileHelper.createFileAtPath(pathToFile)
        Assert.assertEquals(outputFile?.absolutePath, "path/to/parent/filename")
    }


    @Test
    @Throws(IOException::class)
    fun writeToFileShouldReturnTrueIfTextWasWrittenToValidFilePath(@Mocked fileWriter: FileWriter?) {
        val data = "Data"
        val file = File("valid/path/to/file")
        object : Expectations() {
            init {
                FileWriter(file)
                fileWriter!!.write(data)
                fileWriter!!.flush()
                fileWriter!!.close()
            }
        }
        val resultWritingFile = FileHelper.writeToFile(file, data)
        Assert.assertTrue(resultWritingFile)
    }

    @Test
    @Throws(IOException::class)
    fun writeToFileShouldReturnFalseIfTextWasNotWrittenToInvalidFilePath(@Mocked fileWriter: FileWriter?) {
        val data = "Data"
        val file = File("invalid/path/to/file")
        object : Expectations() {
            init {
                FileWriter(file)
                result = IOException()
            }
        }
        val resultWritingFile = FileHelper.writeToFile(file, data)
        Assert.assertFalse(resultWritingFile)
    }

    @Test
    @Throws(IOException::class)
    fun readFileShouldReturnFileContentWhenValidFilePath(@Mocked scanner: Scanner?) {
        val data = "Data"
        val file = File("valid/path/to/file")
        object : Expectations() {
            init {
                Scanner(file)
                scanner!!.useDelimiter("\\Z")
                result = scanner
                scanner!!.next()
                result = data
            }
        }
        Assert.assertEquals(FileHelper.readFile(file), "Data")
    }

    @Test
    @Throws(IOException::class)
    fun readFileShouldReturnNullWhenInvalidFilePath(@Mocked scanner: Scanner?) {
        val data = "Data"
        val file = File("invalid/path/to/file")
        object : Expectations() {
            init {
                Scanner(file)
                result = FileNotFoundException()
            }
        }
        Assert.assertNull(FileHelper.readFile(file))
    }

    @Test
    @Throws(FileNotFoundException::class)
    fun writeBitmapToFileShouldReturnTrueWhenBitmapIsSavedToValidFilePath(
            @Mocked fos: FileOutputStream?,
            @Mocked bitmap: Bitmap?) {
        val file = File("valid/path/to/file")
        object : Expectations() {
            init {
                bitmap!!.compress(Bitmap.CompressFormat.JPEG, 100, FileOutputStream(file))
                result = true
            }
        }
        Assert.assertTrue(FileHelper.writeBitmapToFile(file, bitmap))
    }

    @Test
    @Throws(FileNotFoundException::class)
    fun writeBitmapToFileShouldReturnFalseWhenBitmapIsNotSavedToInvalidFilePath(
            @Mocked fos: FileOutputStream?,
            @Mocked bitmap: Bitmap?) {
        val file = File("invalid/path/to/file")
        object : Expectations() {
            init {
                FileOutputStream(file)
                result = FileNotFoundException()
            }
        }
        Assert.assertFalse(FileHelper.writeBitmapToFile(file, bitmap))
    }

    @Test
    fun writeUriToFileShouldReturnTrueWhenUriIsNotSavedToValidFilePath(
            @Mocked contentResolver: ContentResolver,
            @Mocked fis: FileInputStream) {

        // TODO complete this - cannot seem to mock effectively
    }

    @Test
    fun writeUriToFileShouldReturnTrueWhenUriIsSavedToInvalidFilePath(
            @Mocked contentResolver: ContentResolver,
            @Mocked fis: FileInputStream) {

        // TODO complete this - cannot seem to mock effectively
    }

    @Test
    @Throws(IOException::class)
    fun writeUriToFileShouldReturnFalseWhenMethodThrowsIOException(
            @Mocked contentResolver: ContentResolver,
            @Mocked fis: FileInputStream) {

        // TODO complete this - cannot seem to mock effectively
    }


}